pallets/ds-maps/src/lib.rs

```diff
@@ -0,0 +1,501 @@
+        ZoneCreated(RootId, AreaId, ZoneId, AccountId),
+        /// Area type changed [role, area, root, who]
+        AreaTypeChanged(u8, AreaId, RootId, AccountId),
+        /// Root was removed from storage
+        RootRemoved(RootId, AccountId),
+        /// Zone was removed from storage
+        ZoneRemoved(ZoneId, AccountId),
+    }
+);
+
+// Errors inform users that something went wrong.
+// learn more https://substrate.dev/docs/en/knowledgebase/runtime/errors
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Error names should be descriptive.
+        NoneValue,
+        /// Operation is not valid
+        InvalidAction,
+        /// Incorrect data provided
+        InvalidData,
+        /// Origin do not have sufficient privileges to perform the operation
+        NotAuthorized,
+        /// Account doesn't exist
+        NotExists,
+        /// Area is unavailable for operation
+        ForbiddenArea,
+        /// Root you are trying to access is not in storage
+        RootDoesNotExist,
+        /// Sizes are off bounds
+        BadDimesions,
+        /// Area can't contain no more buildings
+        AreaFull,
+        /// Zone points lies in different areas
+        ZoneDoesntFit,
+        /// Zone you are trying to access is not in storage
+        ZoneDoesntExist,
+        /// Added zone overlaps with another in current area
+        OverlappingZone,
+        // Add additional errors below
+    }
+}
+
+// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+// These functions materialize as "extrinsics", which are often compared to transactions.
+// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Errors must be initialized if they are used by the pallet.
+        type Error = Error<T>;
+
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+        /// Adds new root to storage
+        #[weight = <T as Trait>::WeightInfo::root_add()]
+        pub fn root_add(origin, 
+                        bounding_box: Box3D<T::Coord>,
+                        delta: T::Coord) -> dispatch::DispatchResult {
+            let who = ensure_signed(origin)?;
+            ensure!(<accounts::Module<T>>::account_is(&who, REGISTRAR_ROLE.into()), Error::<T>::NotAuthorized);
+            // TODO replace these ensures w inverted index (using global grid)
+            let root_size = bounding_box.projection_on_plane().get_dimensions();
+            ensure!(root_size.lat <= Self::coord_from_str("1"), Error::<T>::BadDimesions);
+            ensure!(root_size.lon <= Self::coord_from_str("1"), Error::<T>::BadDimesions);
+            ensure!(delta <= Self::coord_from_str("0.1") && 
+                    delta >= Self::coord_from_str("0.002"), Error::<T>::InvalidData);
```

Comment:
I'd strongly suggest to move these "magic numbers" to constants or pallet parameters, and write corresponding comments there - why specifically these numbers, how they were calculated.